rfcs/0189-contracts.md

```diff
@@ -0,0 +1,658 @@
+in the expected format by the upstream service.
+This moves the possible failure at evaluation time which is a very nice property.
+
+[kadmin]: https://github.com/NixOS/nixpkgs/blob/nixos-25.05/nixos/modules/services/security/kanidm.nix
+
+_Aside: This is such big step forward in user experience that we would like_
+_to see this more available. This will be tackled though in the [vars][] proposal_
+_and **not** in this RFC. The `vars` proposal will use the secrets contract_
+_as presented here or a slightly modified if deemed necessary._
+
+[vars]: https://discourse.nixos.org/t/vars-a-framework-for-managing-secrets-and-computed-values/62411
+
+One problem encountered by those modules providing one option per secret
+is the file must be readable by the user of the service.
+This is often solved by relying on [systemd's credentials][] system
+or less securely by using the `root` user in the service startup to read from the file.
+
+[systemd's credentials]: https://systemd.io/CREDENTIALS/
+
+This contract provides an alternative where the `consumer` of the contract - the module requiring a secret - imposes a `user` to the secret `provider`, which here would be [agenix][] or [sops-nix][] for example.
+
+[agenix]: https://github.com/ryantm/agenix
+[sops-nix]: https://github.com/Mic92/sops-nix
+
+Contrary to the previous contracts we covered, the `consumer` here needs to read the `output` of the `provider`
+because it contains the path to the file containing the secret.
+
+When testing a module that expects a file containing a raw secret,
+the ubiquitous method to provide the file is using `pkgs.writeText`.
+This works but has the issue the created file is world readable
+and we thus do not test the file is accessible with the correct user.
+To avoid this pitfall going forwards, we created the [`testing.hardcodedSecret`
+`provider`][hardcodedSecret: new secret contract consumer]
+which is an improved version of `pkgs.writeText`
+where the resulting file is created with the requested `owner`, `mode`, etc.
+as described by the contract `consumer`.
+
+[hardcodedSecret: new secret contract consumer]: https://github.com/NixOS/nixpkgs/pull/432529/commits/6fbd099aa306d2cce337b8fa7ed7e0c8a255aebf
+
+This new provider has been tested using the [contract's behavior test][hardcodedSecret: define behavior test for secret contract]
+and has been used in [`services.stash`'s module][stash: use secret contract for passwords] as an example.
+
+[hardcodedSecret: define behavior test for secret contract]: https://github.com/NixOS/nixpkgs/pull/432529/commits/448410a520225bc71e1616611cef7ad086c64cd1
+[stash: use secret contract for passwords]: https://github.com/NixOS/nixpkgs/pull/432529/commits/19419ad95913fbed4636d0b24d95c80517c18340
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+We are not aware of any because this solution is fully backwards compatible,
+incremental and has a lot advantages. It also arose from a real practical need.
```

Comment:
Standard question for any module-system-based increase in layers: what is the estimated evaluation-time resource impact?

Reply:
Good question; I don't actually know things beyond `export NIX_SHOW_STATS=1`. There are some scripts in `ci/eval` but I am not sure whether reusing parts of them is worth it (but maybe looking at what is done there could be useful)